src/ucs/memory/numa.c

```diff
@@ -72,3 +105,242 @@ int ucs_numa_node_of_cpu(int cpu)
 }
 
 #endif
+
+static inline uint32_t ucs_numa_distance_hash_key(const ucs_numa_node_t node1,
+                                                  const ucs_numa_node_t node2)
+{
+    static const uint8_t shift = sizeof(ucs_numa_node_t) * 8;
+
+    return (ucs_max(node1, node2) << shift) | ucs_min(node1, node2);
+}
+
+/**
+ * Iterate the directory entries in @param path of the form
+ * @param prefix<index> and return the maximum index (positive number).
+ *
+ * @param [in]  path          Directory path
+ * @param [in]  prefix        Consider only entries starting with this prefix
+ * @param [in]  prefix_length Prefix length
+ *
+ * @return Maximum index of the entries starting with prefix or -1 on error.
+ */
+static int ucs_numa_get_max_dirent(const char *path, const char *prefix,
+                                   const size_t prefix_length)
+{
+    int max_index = 0;
+    DIR *dir;
+    struct dirent *de;
+    int idx;
+
+    dir = opendir(path);
+    if (!dir) {
+        return -1;
+    } else {
+        while ((de = readdir(dir)) != NULL) {
+            if (strncmp(de->d_name, prefix, prefix_length)) {
+                continue;
+            }
+
+            idx = strtoul(de->d_name + prefix_length, 0, 0);
+            if (max_index < idx) {
+                max_index = idx;
+            }
+        }
+
+        closedir(dir);
+    }
+
+    return max_index;
+}
+
+int ucs_numa_num_configured_nodes()
+{
+    static int max_node = -1;
+
+    if (max_node == -1) {
+        UCS_NUMA_GET_MAX_DIRENT(max_node, UCS_NUMA_NODES_DIR_PATH, "node",
+                                UCS_NUMA_MAX_NODE, UCS_NUMA_DEFAULT_NODE);
+    }
+
+    return max_node;
+}
+
+int ucs_numa_num_configured_cpus()
+{
+    static int max_cpu = -1;
+
+    ucs_spin_lock(&ucs_numa_global_ctx.lock);
+    if (max_cpu == -1) {
+        UCS_NUMA_GET_MAX_DIRENT(max_cpu, UCS_NUMA_CORES_DIR_PATH, "cpu",
+                                (__CPU_SETSIZE - 1), 0);
+        max_cpu += 1;
+    }
+
+    ucs_spin_unlock(&ucs_numa_global_ctx.lock);
+    return max_cpu;
+}
+
+ucs_numa_node_t ucs_numa_node_of_cpu_v2(int cpu)
+{
+    /* Used for caching to improve perfromance */
+    static ucs_numa_node_t cpu_numa_node[__CPU_SETSIZE] = {0};
+    int node                                            = 0;
+    char core_dir_path[PATH_MAX];
+
+    ucs_assert(cpu < __CPU_SETSIZE);
+
+    ucs_spin_lock(&ucs_numa_global_ctx.lock);
+    if (cpu_numa_node[cpu] == 0) {
+        ucs_snprintf_safe(core_dir_path, PATH_MAX, UCS_NUMA_CORE_DIR_PATH, cpu);
+        UCS_NUMA_GET_MAX_DIRENT(node, core_dir_path, "node",
+                                ucs_numa_num_configured_nodes(),
+                                UCS_NUMA_DEFAULT_NODE);
+        cpu_numa_node[cpu] = node + 1;
+    }
+
+    ucs_spin_unlock(&ucs_numa_global_ctx.lock);
+    return cpu_numa_node[cpu] - 1;
+}
+
+ucs_numa_node_t ucs_numa_node_of_device(const char *dev_path)
```

Comment:
`of` adds some clumsiness to the naming.
WDYT of:
`ucs_numa_(get)_device_node`?

Reply:
Just followed anming convention of `ucs_numa_node_of_cpu`
I'm guessing it's coming from libnuma API, it has function named `numa_node_of_cpu`